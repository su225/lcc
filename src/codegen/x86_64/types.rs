use std::fmt::{Display, Formatter};
use crate::codegen::x86_64::register::Register;
use crate::tacky::TackySymbol;

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct StackOffset(pub(crate) isize);

#[derive(Debug, Clone, PartialEq)]
pub enum AsmOperand {
    Imm32(i32),
    Reg(Register),
    Pseudo(TackySymbol),
    Stack { offset: StackOffset },
}

impl Display for AsmOperand {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", match &self {
            AsmOperand::Imm32(n) => format!("${}", n),
            AsmOperand::Reg(r) => r.to_string(),
            AsmOperand::Pseudo(p) => format!("<<{}>>", p),
            AsmOperand::Stack { offset } => format!("{}(%rbp)", offset.0),
        })
    }
}

#[derive(Debug, PartialEq)]
pub enum AsmInstruction {
    AllocateStack(usize),
    Mov8 { src: AsmOperand, dst: AsmOperand },
    Mov32 { src: AsmOperand, dst: AsmOperand },
    Neg32 { op: AsmOperand },
    Not32 { op: AsmOperand },
    Add32 { src: AsmOperand, dst: AsmOperand },
    Sub32 { src: AsmOperand, dst: AsmOperand },
    IMul32 { src: AsmOperand, dst: AsmOperand },
    IDiv32 { divisor: AsmOperand },
    And32 { src: AsmOperand, dst: AsmOperand },
    Or32 { src: AsmOperand, dst: AsmOperand },
    Xor32 { src: AsmOperand, dst: AsmOperand },
    Sal32 { src: AsmOperand, dst: AsmOperand },
    Sar32 { src: AsmOperand, dst: AsmOperand },
    Shl32 { src: AsmOperand, dst: AsmOperand },
    Shr32 { src: AsmOperand, dst: AsmOperand },
    SignExtendTo64, // Sign extend %eax to 64-bit into %edx
    Ret,
}

#[derive(Debug, PartialEq)]
pub struct AsmFunction {
    pub name: TackySymbol,
    pub instructions: Vec<AsmInstruction>,
}

#[derive(Debug, PartialEq)]
pub struct AsmProgram {
    pub functions: Vec<AsmFunction>,
}