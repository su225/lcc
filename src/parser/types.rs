use derive_more::Add;
use serde::Serialize;
use crate::common::{Location, Radix};

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub struct Symbol<'a> {
    pub name: &'a str,
    pub(crate) location: Location,
}

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum UnaryOperator {
    Complement,
    Negate,
    Not,
}

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum BinaryOperatorAssociativity {
    Left, Right
}

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum BinaryOperator {
    Add,
    Subtract,
    Multiply,
    Divide,
    Modulo,

    BitwiseAnd,
    BitwiseOr,
    BitwiseXor,
    LeftShift,
    RightShift,

    And,
    Or,
    Equal,
    NotEqual,
    LessThan,
    LessThanOrEqual,
    GreaterThan,
    GreaterThanOrEqual,
}

#[derive(Debug, PartialEq, Ord, PartialOrd, Eq, Add, Serialize)]
#[serde(rename_all = "snake_case")]
pub struct BinaryOperatorPrecedence(pub(crate) u16);

impl BinaryOperator {
    #[inline]
    pub(crate) fn associativity(&self) -> BinaryOperatorAssociativity {
        match self {
            BinaryOperator::Add => BinaryOperatorAssociativity::Left,
            BinaryOperator::Subtract => BinaryOperatorAssociativity::Left,
            BinaryOperator::Multiply => BinaryOperatorAssociativity::Left,
            BinaryOperator::Divide => BinaryOperatorAssociativity::Left,
            BinaryOperator::Modulo => BinaryOperatorAssociativity::Left,
            BinaryOperator::BitwiseAnd => BinaryOperatorAssociativity::Left,
            BinaryOperator::BitwiseOr => BinaryOperatorAssociativity::Left,
            BinaryOperator::BitwiseXor => BinaryOperatorAssociativity::Left,
            BinaryOperator::LeftShift => BinaryOperatorAssociativity::Left,
            BinaryOperator::RightShift => BinaryOperatorAssociativity::Left,
            BinaryOperator::And => BinaryOperatorAssociativity::Left,
            BinaryOperator::Or => BinaryOperatorAssociativity::Left,
            BinaryOperator::Equal => BinaryOperatorAssociativity::Left,
            BinaryOperator::NotEqual => BinaryOperatorAssociativity::Left,
            BinaryOperator::LessThan => BinaryOperatorAssociativity::Left,
            BinaryOperator::LessThanOrEqual => BinaryOperatorAssociativity::Left,
            BinaryOperator::GreaterThan => BinaryOperatorAssociativity::Left,
            BinaryOperator::GreaterThanOrEqual => BinaryOperatorAssociativity::Left,
        }
    }

    #[inline]
    pub(crate) fn precedence(&self) -> BinaryOperatorPrecedence {
        match self {
            BinaryOperator::Multiply
            | BinaryOperator::Divide
            | BinaryOperator::Modulo => BinaryOperatorPrecedence(50),

            BinaryOperator::Add | BinaryOperator::Subtract => BinaryOperatorPrecedence(45),
            BinaryOperator::LeftShift | BinaryOperator::RightShift => BinaryOperatorPrecedence(42),

            BinaryOperator::LessThan
            | BinaryOperator::LessThanOrEqual
            | BinaryOperator::GreaterThan
            | BinaryOperator::GreaterThanOrEqual => BinaryOperatorPrecedence(40),

            BinaryOperator::Equal | BinaryOperator::NotEqual => BinaryOperatorPrecedence(38),

            BinaryOperator::BitwiseAnd => BinaryOperatorPrecedence(36),
            BinaryOperator::BitwiseXor => BinaryOperatorPrecedence(34),
            BinaryOperator::BitwiseOr => BinaryOperatorPrecedence(32),

            BinaryOperator::And => BinaryOperatorPrecedence(30),
            BinaryOperator::Or => BinaryOperatorPrecedence(28),
        }
    }
}

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum ExpressionKind<'a> {
    IntConstant(&'a str, Radix),
    Unary(UnaryOperator, Box<Expression<'a>>),
    Binary(BinaryOperator, Box<Expression<'a>>, Box<Expression<'a>>),
}

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub struct Expression<'a> {
    pub(crate) location: Location,
    pub kind: ExpressionKind<'a>,
}

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum PrimitiveKind {
    Integer,
    UnsignedInteger,
    LongInteger,
}

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum TypeExpressionKind {
    Primitive(PrimitiveKind),
}

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub struct TypeExpression {
    pub location: Location,
    pub kind: TypeExpressionKind,
}

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum StatementKind<'a> {
    Return(Expression<'a>),
}

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub struct Statement<'a> {
    pub location: Location,
    pub kind: StatementKind<'a>,
}

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub struct FunctionDefinition<'a> {
    pub location: Location,
    pub name: Symbol<'a>,
    pub body: Vec<Statement<'a>>,
}

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub struct ProgramDefinition<'a> {
    pub functions: Vec<FunctionDefinition<'a>>,
}