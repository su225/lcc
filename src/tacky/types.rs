use derive_more::Display;
use crate::parser::types::{BinaryOperator, UnaryOperator};

pub(crate) const COMPILER_GEN_PREFIX: &'static str = "<t>";

#[derive(Debug, Eq, PartialEq, Clone, Hash, Display)]
pub struct TackySymbol(pub String);

impl From<&str> for TackySymbol {
    fn from(value: &str) -> Self {
        TackySymbol(value.to_string())
    }
}

impl TackySymbol {
    pub fn is_generated(&self) -> bool {
        self.0.starts_with(COMPILER_GEN_PREFIX)
    }
}

#[derive(Debug, PartialEq)]
pub struct TackyProgram {
    pub functions: Vec<TackyFunction>,
}

#[derive(Debug, PartialEq)]
pub struct TackyFunction {
    pub identifier: TackySymbol,
    pub body: Vec<TackyInstruction>,
}

#[derive(Debug, PartialEq)]
pub enum TackyUnaryOperator {
    Complement,
    Negate,
    Not,
}

impl From<&UnaryOperator> for TackyUnaryOperator {
    fn from(value: &UnaryOperator) -> Self {
        match value {
            UnaryOperator::Complement => TackyUnaryOperator::Complement,
            UnaryOperator::Negate => TackyUnaryOperator::Negate,
            UnaryOperator::Not => todo!(),
        }
    }
}

#[derive(Debug, PartialEq)]
pub enum TackyBinaryOperator {
    Add,
    Subtract,
    Multiply,
    Divide,
    Modulo,
    BitwiseAnd,
    BitwiseOr,
    BitwiseXor,
    LeftShift,
    RightShift,
    Equal,
    NotEqual,
    LessThan,
    LessOrEqual,
    GreaterThan,
    GreaterOrEqual,
}

impl From<&BinaryOperator> for TackyBinaryOperator {
    fn from(value: &BinaryOperator) -> Self {
        match value {
            BinaryOperator::Add => TackyBinaryOperator::Add,
            BinaryOperator::Subtract => TackyBinaryOperator::Subtract,
            BinaryOperator::Multiply => TackyBinaryOperator::Multiply,
            BinaryOperator::Divide => TackyBinaryOperator::Divide,
            BinaryOperator::Modulo => TackyBinaryOperator::Modulo,

            BinaryOperator::BitwiseAnd => TackyBinaryOperator::BitwiseAnd,
            BinaryOperator::BitwiseOr => TackyBinaryOperator::BitwiseOr,
            BinaryOperator::BitwiseXor => TackyBinaryOperator::BitwiseXor,
            BinaryOperator::LeftShift => TackyBinaryOperator::LeftShift,
            BinaryOperator::RightShift => TackyBinaryOperator::RightShift,

            _ => todo!(),
        }
    }
}

#[derive(Debug, PartialEq)]
pub enum TackyValue {
    Constant32(i32),
    Variable(TackySymbol),
}

#[derive(Debug, PartialEq)]
pub enum TackyInstruction {
    Unary {
        operator: TackyUnaryOperator,
        src: TackyValue,
        dst: TackyValue,
    },
    Binary {
        operator: TackyBinaryOperator,
        src1: TackyValue,
        src2: TackyValue,
        dst: TackyValue,
    },
    Return(TackyValue),
    Copy { src: TackyValue, dst: TackyValue },
    Jump { target: TackySymbol },
    JumpIfZero { condition: TackyValue, target: TackySymbol },
    JumpIfNotZero { condition: TackyValue, target: TackySymbol },
    Label(TackySymbol),
}