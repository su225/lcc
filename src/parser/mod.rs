pub mod parser;
mod types;
mod errors;

pub use parser::*;