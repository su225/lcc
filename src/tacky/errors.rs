use std::num::ParseIntError;
use thiserror::Error;

#[derive(Error, PartialEq, Debug)]
pub enum TackyError {
    #[error(transparent)]
    IntImmediateParseError(#[from] ParseIntError),
}