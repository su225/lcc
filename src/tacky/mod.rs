pub mod tacky;

pub use tacky::*;