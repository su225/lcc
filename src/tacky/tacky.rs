use std::fmt::{Display, Formatter};
use std::num::ParseIntError;
use derive_more::Display;
use thiserror::Error;
use crate::parser::types::*;
use crate::tacky::TackyInstruction::*;
use crate::tacky::TackyValue::*;

pub(crate) const COMPILER_GEN_PREFIX: &'static str = "<t>";

#[derive(Debug, Eq, PartialEq, Clone, Hash, Display)]
pub struct TackySymbol(pub String);

impl From<&str> for TackySymbol {
    fn from(value: &str) -> Self {
        TackySymbol(value.to_string())
    }
}

impl TackySymbol {
    pub fn is_generated(&self) -> bool {
        self.0.starts_with(COMPILER_GEN_PREFIX)
    }
}

#[derive(Debug, PartialEq)]
pub struct TackyProgram {
    pub functions: Vec<TackyFunction>,
}

impl Display for TackyProgram {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        for tacky_func in self.functions.iter() {
            tacky_func.fmt(f)?;
        }
        Ok(())
    }
}

#[derive(Debug, PartialEq)]
pub struct TackyFunction {
    pub identifier: TackySymbol,
    pub body: Vec<TackyInstruction>,
}

impl Display for TackyFunction {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.write_fmt(format_args!("int {}(void) {{\n", self.identifier.0))?;
        for instr in self.body.iter() {
            f.write_fmt(format_args!("{}\n", instr))?;
        }
        f.write_str("}\n\n")?;
        Ok(())
    }
}

#[derive(Debug, PartialEq)]
pub enum TackyUnaryOperator {
    Complement,
    Negate,
    Not,
}

impl From<&UnaryOperator> for TackyUnaryOperator {
    fn from(value: &UnaryOperator) -> Self {
        match value {
            UnaryOperator::Complement => TackyUnaryOperator::Complement,
            UnaryOperator::Negate => TackyUnaryOperator::Negate,
            UnaryOperator::Not => TackyUnaryOperator::Not,
        }
    }
}

impl Display for TackyUnaryOperator {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            TackyUnaryOperator::Complement => f.write_str("~"),
            TackyUnaryOperator::Negate => f.write_str("-"),
            TackyUnaryOperator::Not => f.write_str("!"),
        }
    }
}

#[derive(Debug, PartialEq)]
pub enum TackyBinaryOperator {
    Add,
    Subtract,
    Multiply,
    Divide,
    Modulo,
    BitwiseAnd,
    BitwiseOr,
    BitwiseXor,
    LeftShift,
    RightShift,
    Equal,
    NotEqual,
    LessThan,
    LessOrEqual,
    GreaterThan,
    GreaterOrEqual,
}

impl Display for TackyBinaryOperator {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            TackyBinaryOperator::Add => f.write_str("+"),
            TackyBinaryOperator::Subtract => f.write_str("-"),
            TackyBinaryOperator::Multiply => f.write_str("*"),
            TackyBinaryOperator::Divide => f.write_str("/"),
            TackyBinaryOperator::Modulo => f.write_str("%"),
            TackyBinaryOperator::BitwiseAnd => f.write_str("&"),
            TackyBinaryOperator::BitwiseOr => f.write_str("|"),
            TackyBinaryOperator::BitwiseXor => f.write_str("^"),
            TackyBinaryOperator::LeftShift => f.write_str("<<"),
            TackyBinaryOperator::RightShift => f.write_str(">>"),
            TackyBinaryOperator::Equal => f.write_str("=="),
            TackyBinaryOperator::NotEqual => f.write_str("!="),
            TackyBinaryOperator::LessThan => f.write_str("<"),
            TackyBinaryOperator::LessOrEqual => f.write_str("<="),
            TackyBinaryOperator::GreaterThan => f.write_str(">"),
            TackyBinaryOperator::GreaterOrEqual => f.write_str(">="),
        }
    }
}

impl From<&BinaryOperator> for TackyBinaryOperator {
    fn from(value: &BinaryOperator) -> Self {
        match value {
            BinaryOperator::Add => TackyBinaryOperator::Add,
            BinaryOperator::Subtract => TackyBinaryOperator::Subtract,
            BinaryOperator::Multiply => TackyBinaryOperator::Multiply,
            BinaryOperator::Divide => TackyBinaryOperator::Divide,
            BinaryOperator::Modulo => TackyBinaryOperator::Modulo,

            BinaryOperator::BitwiseAnd => TackyBinaryOperator::BitwiseAnd,
            BinaryOperator::BitwiseOr => TackyBinaryOperator::BitwiseOr,
            BinaryOperator::BitwiseXor => TackyBinaryOperator::BitwiseXor,
            BinaryOperator::LeftShift => TackyBinaryOperator::LeftShift,
            BinaryOperator::RightShift => TackyBinaryOperator::RightShift,

            BinaryOperator::And => panic!("logical AND cannot be translated as binary operator"),
            BinaryOperator::Or => panic!("logical OR cannot be translated as binary operator"),
            BinaryOperator::Equal => TackyBinaryOperator::Equal,
            BinaryOperator::NotEqual => TackyBinaryOperator::NotEqual,
            BinaryOperator::LessThan => TackyBinaryOperator::LessThan,
            BinaryOperator::LessThanOrEqual => TackyBinaryOperator::LessOrEqual,
            BinaryOperator::GreaterThan => TackyBinaryOperator::GreaterThan,
            BinaryOperator::GreaterThanOrEqual => TackyBinaryOperator::GreaterOrEqual,
        }
    }
}

#[derive(Debug, PartialEq)]
pub enum TackyValue {
    Int32(i32),
    Variable(TackySymbol),
}

impl Display for TackyValue {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            Int32(c) => f.write_fmt(format_args!("(int32 {})", c)),
            Variable(var) => f.write_str(&(var.0.clone())),
        }
    }
}

#[derive(Debug, PartialEq)]
pub enum TackyInstruction {
    Unary {
        operator: TackyUnaryOperator,
        src: TackyValue,
        dst: TackyValue,
    },
    Binary {
        operator: TackyBinaryOperator,
        src1: TackyValue,
        src2: TackyValue,
        dst: TackyValue,
    },
    Return(TackyValue),
    Copy { src: TackyValue, dst: TackyValue },
    Jump { target: TackySymbol },
    JumpIfZero { condition: TackyValue, target: TackySymbol },
    JumpIfNotZero { condition: TackyValue, target: TackySymbol },
    Label(TackySymbol),
}

impl Display for TackyInstruction {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            Unary { operator, src, dst } => {
                f.write_fmt(format_args!("    {dst} = {operator} {src};"))
            },
            Binary { operator, src1, src2, dst } => {
                f.write_fmt(format_args!("    {dst} = {src1} {operator} {src2};"))
            },
            Return(ret_value) => {
                f.write_fmt(format_args!("    return {ret_value};"))
            },
            Copy { src, dst } => {
                f.write_fmt(format_args!("    {src} = {dst};"))
            },
            Jump { target } => {
                f.write_fmt(format_args!("    jump {target};"))
            },
            JumpIfZero { condition, target } => {
                f.write_fmt(format_args!("    jump_if_zero {condition} {target};"))
            }
            JumpIfNotZero { condition, target } => {
                f.write_fmt(format_args!("    jump_if_not_zero {condition} {target};"))
            },
            Label(lbl) => f.write_fmt(format_args!("{}:", lbl.0)),
        }
    }
}

#[derive(Error, PartialEq, Debug)]
pub enum TackyError {
    #[error(transparent)]
    IntImmediateParseError(#[from] ParseIntError),
}

struct TackyContext {
    next_int: i64,
}

impl TackyContext {
    fn new() -> TackyContext {
        TackyContext { next_int: 0 }
    }

    fn next_temporary_identifier(&mut self) -> TackySymbol {
        let identifier = format!("{}.{}", COMPILER_GEN_PREFIX, self.next_int);
        self.next_int += 1;
        TackySymbol(identifier)
    }
}

pub fn emit(prog: &ProgramDefinition) -> Result<TackyProgram, TackyError> {
    let mut ctx = TackyContext::new();
    let mut f = vec![];
    for fd in prog.functions.iter() {
        let tf = emit_tacky_for_function(&mut ctx, fd)?;
        f.push(tf);
    }
    Ok(TackyProgram { functions: f })
}

fn emit_tacky_for_function(ctx: &mut TackyContext, f: &FunctionDefinition) -> Result<TackyFunction, TackyError> {
    let mut instructions = vec![];
    for stmt in f.body.iter() {
        let instrs = emit_tacky_for_statement(ctx, stmt)?;
        instructions.extend(instrs);
    }
    Ok(TackyFunction {
        identifier: TackySymbol(f.name.name.into()),
        body: instructions,
    })
}

fn emit_tacky_for_statement(ctx: &mut TackyContext, s: &Statement) -> Result<Vec<TackyInstruction>, TackyError> {
    match s.kind {
        StatementKind::Return(ref expr) => {
            let (dst, mut expr_instrs) = emit_tacky_for_expression(ctx, expr)?;
            expr_instrs.push(Return(dst));
            Ok(expr_instrs)
        }
    }
}

fn emit_tacky_for_expression(ctx: &mut TackyContext, e: &Expression) -> Result<(TackyValue, Vec<TackyInstruction>), TackyError> {
    match &e.kind {
        ExpressionKind::IntConstant(c, radix) => {
            let n = i32::from_str_radix(c, radix.value())?;
            Ok((Int32(n), vec![]))
        }
        ExpressionKind::Unary(unary_op, src) => {
            let (src_tacky, mut tacky_instrs) = emit_tacky_for_expression(ctx, src)?;
            let dst_tacky_identifier = ctx.next_temporary_identifier();
            let dst_tacky = Variable(dst_tacky_identifier.clone());
            let result_val = Variable(dst_tacky_identifier);
            tacky_instrs.push(Unary {
                operator: TackyUnaryOperator::from(unary_op),
                src: src_tacky,
                dst: dst_tacky,
            });
            Ok((result_val, tacky_instrs))
        }
        ExpressionKind::Binary(binary_op, op1, op2) => {
            let (src1_tacky, mut tacky_instrs) = emit_tacky_for_expression(ctx, op1)?;
            let (src2_tacky, src2_tacky_instrs) = emit_tacky_for_expression(ctx, op2)?;
            let dst_tacky_identifier = ctx.next_temporary_identifier();
            let dst_tacky = Variable(dst_tacky_identifier.clone());
            let result = Variable(dst_tacky_identifier);
            tacky_instrs.extend(src2_tacky_instrs);
            tacky_instrs.push(Binary {
                operator: TackyBinaryOperator::from(binary_op),
                src1: src1_tacky,
                src2: src2_tacky,
                dst: dst_tacky,
            });
            Ok((result, tacky_instrs))
        }
    }
}

#[cfg(test)]
mod test {
    use ExpressionKind::*;
    use crate::common::Radix;
    use crate::parser::types::{Expression, ExpressionKind, UnaryOperator};
    use crate::tacky::tacky::{emit_tacky_for_expression, TackyContext};
    use crate::tacky::*;
    use crate::tacky::TackyValue::*;

    #[test]
    fn test_emit_tacky_for_int_constant() {
        let expr = Expression {
            location: (0,0).into(),
            kind: IntConstant("10", Radix::Decimal),
        };
        let mut ctx = TackyContext::new();
        let (tval, tinstrs) = emit_tacky_for_expression(&mut ctx, &expr)
            .expect("error while generating IR");
        assert_eq!(tval, Int32(10));
        assert_eq!(tinstrs, vec![]);
    }

    #[test]
    fn test_emit_tacky_for_unary_complement() {
        let expr = Expression {
            location: (0,0).into(),
            kind: Unary(UnaryOperator::Complement,
                Box::new(Expression {
                    location: (0,0).into(),
                    kind: IntConstant("10", Radix::Decimal),
                })),
        };
        let mut ctx = TackyContext::new();
        let (tval, tinstrs) = emit_tacky_for_expression(&mut ctx, &expr).unwrap();
        assert_eq!(tval, Variable(TackySymbol("<t>.0".into())));
        assert_eq!(tinstrs, vec![
            TackyInstruction::Unary {
                operator: TackyUnaryOperator::Complement,
                src: Int32(10),
                dst: Variable(TackySymbol("<t>.0".into())),
            },
        ]);
    }

    #[test]
    fn test_emit_tacky_for_unary_negate() {
        let expr = Expression {
            location: (0,0).into(),
            kind: Unary(UnaryOperator::Negate,
                        Box::new(Expression {
                            location: (0,0).into(),
                            kind: IntConstant("10", Radix::Decimal),
                        })),
        };
        let mut ctx = TackyContext::new();
        let (tval, tinstrs) = emit_tacky_for_expression(&mut ctx, &expr).unwrap();
        assert_eq!(tval, Variable(TackySymbol("<t>.0".into())));
        assert_eq!(tinstrs, vec![
            TackyInstruction::Unary {
                operator: TackyUnaryOperator::Negate,
                src: Int32(10),
                dst: Variable(TackySymbol("<t>.0".into())),
            },
        ]);
    }

    #[test]
    fn test_emit_tacky_for_unary_not() {
        let expr = Expression {
            location: (0,0).into(),
            kind: Unary(UnaryOperator::Not,
                        Box::new(Expression {
                            location: (0,0).into(),
                            kind: IntConstant("10", Radix::Decimal),
                        })),
        };
        let mut ctx = TackyContext::new();
        let (tval, tinstrs) = emit_tacky_for_expression(&mut ctx, &expr).unwrap();
        assert_eq!(tval, Variable(TackySymbol("<t>.0".into())));
        assert_eq!(tinstrs, vec![
            TackyInstruction::Unary {
                operator: TackyUnaryOperator::Not,
                src: Int32(10),
                dst: Variable(TackySymbol("<t>.0".into())),
            },
        ]);
    }
}

#[cfg(test)]
mod tacky_ir_generation_snapshot_tests {
    use std::fs;
    use std::path::{Path, PathBuf};
    use insta::assert_snapshot;
    use rstest::rstest;
    use crate::lexer::Lexer;
    use crate::parser::Parser;
    use crate::tacky::emit;

    #[rstest]
    #[case("multifunc/simple.c")]
    fn test_generation_for_multiple_functions(#[case] input_path: &str) {
        run_ir_generation_snapshot_test("multiple functions", input_path);
    }

    #[rstest]
    #[case("unary/complement.c")]
    #[case("unary/negation.c")]
    #[case("unary/not.c")]
    fn test_generation_for_unary_operators(#[case] input_path: &str) {
        run_ir_generation_snapshot_test("unary operators", input_path)
    }

    #[rstest]
    #[case("binary/arithmetic_add.c")]
    #[case("binary/arithmetic_add_leftassoc.c")]
    #[case("binary/arithmetic_precedence.c")]
    #[case("binary/arithmetic_precedence_override.c")]
    #[case("binary/bitwise_and.c")]
    #[case("binary/relational_eq.c")]
    fn test_generation_for_binary_operators(#[case] input_path: &str) {
        run_ir_generation_snapshot_test("binary operators", input_path)
    }

    fn run_ir_generation_snapshot_test(suite_description: &str, src_file: &str) {
        let base_dir = file!();
        let src_path = Path::new(base_dir).parent().unwrap().join("input").join(src_file);
        let source = fs::read_to_string(src_path.clone());
        assert!(source.is_ok(), "failed to read {:?}", src_path);

        let src = source.unwrap();
        let lexer = Lexer::new(&src);
        let mut parser = Parser::new(lexer);
        let ast = parser.parse().unwrap();
        let tacky = emit(&ast).unwrap();

        let (out_dir, snapshot_file) = output_path_parts(src_file);
        insta::with_settings!({
            sort_maps => true,
            prepend_module_to_snapshot => false,
            description => suite_description,
            snapshot_path => out_dir,
            info => &format!("{}", src_file),
        }, {
            assert_snapshot!(snapshot_file, tacky);
        });
    }

    fn output_path_parts(src_file: &str) -> (PathBuf, String) {
        let input_path = Path::new(src_file);
        let parent = input_path.parent().unwrap_or_else(|| Path::new(""));
        let stem = input_path.file_stem().expect("No file stem").to_string_lossy();
        let output_dir = Path::new("output").join(parent);
        let output_file = format!("{}.tacky", stem);
        (output_dir, output_file)
    }
}