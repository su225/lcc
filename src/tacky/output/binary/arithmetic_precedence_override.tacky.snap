---
source: src/tacky/tacky.rs
description: binary operators
expression: tacky
info: binary/arithmetic_precedence_override.c
---
int main(void) {
    <t>.0 = (int32 1) + (int32 3);
    <t>.1 = (int32 2) * <t>.0;
    return <t>.1;
}
